compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1880,21 +1881,103 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) -> DiagnosticBuilder<'_> {
-        type_error_struct!(
+        debug!("no_such_field_err(span: {:?}, field: {:?}, expr_t: {:?})", span, field, expr_t);
+
+        let mut err = type_error_struct!(
             self.tcx().sess,
             span,
             expr_t,
             E0609,
             "no field `{}` on type `{}`",
             field,
             expr_t
-        )
+        );
+
+        // try to add a suggestion in case the field is a nested field of a field of the Adt
+        if let Some((fields, substs)) = self.get_field_candidates(span, &expr_t) {
+            for candidate_field in fields.iter() {
+                if let Some(field_path) =
+                    self.check_for_nested_field(span, field, candidate_field, substs, vec![])
+                {
+                    let base_sp = base_span.with_hi(field.span.lo());
+                    let mut suggestion_str =
+                        self.tcx.sess.source_map().span_to_snippet(base_sp).unwrap();
+                    let field_path_str = field_path
+                        .iter()
+                        .map(|id| id.name.to_ident_string())
+                        .collect::<Vec<String>>()
+                        .join(".");
+                    debug!("field_path_str: {:?}", field_path_str);
+                    suggestion_str.push_str(&field_path_str);
+
+                    err.span_suggestion(
+                        span,
+                        "try to use the whole path to the field",
+                        suggestion_str,
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+            }
+        }
+        err
+    }
+
+    fn get_field_candidates(
+        &self,
+        span: Span,
+        base_t: Ty<'tcx>,
+    ) -> Option<(&Vec<ty::FieldDef>, SubstsRef<'tcx>)> {
+        let mut autoderef = self.autoderef(span, base_t);
+        while let Some((base_t, _)) = autoderef.next() {
+            match base_t.kind() {
+                ty::Adt(base_def, substs) if !base_def.is_enum() => {
+                    let fields = &base_def.non_enum_variant().fields;
+                    return Some((fields, substs));
+                }
+                _ => {}
+            }
+        }
+        None
+    }
+
+    fn check_for_nested_field(
+        &self,
+        span: Span,
+        target_field: Ident,
+        candidate_field: &ty::FieldDef,
+        subst: SubstsRef<'tcx>,
+        field_path: Vec<Ident>,
+    ) -> Option<Vec<Ident>> {
+        let field_ty = candidate_field.ty(self.tcx, subst);
+        if let Some((nested_fields, _)) = self.get_field_candidates(span, &field_ty) {
+            for field in nested_fields.iter() {
+                let mut field_path_clone = field_path.clone();
+                let ident = field.ident.normalize_to_macros_2_0();
+                if ident == target_field {
+                    field_path_clone.push(ident);
+                    return Some(field_path_clone);
```

Comment:
If we change the suggestion to _not_ include the existing ident, so that the underline in the suggestion only marks the _new_ code, then for the "simple" case of one level nesting you'd return `Some(vec![])`.